packages/vm/lib/runBlock.ts

```diff
@@ -221,12 +239,30 @@ async function applyTransactions(this: VM, block: any, opts: RunBlockOpts) {
     // Combine blooms via bitwise OR
     bloom.or(txRes.bloom)
 
-    const txReceipt: TxReceipt = {
-      status: txRes.execResult.exceptionError ? 0 : 1, // Receipts have a 0 as status on error
+    const abstractTxReceipt: TxReceipt = {
       gasUsed: gasUsed.toArrayLike(Buffer),
       bitvector: txRes.bloom.bitvector,
       logs: txRes.execResult.logs || [],
     }
+    let txReceipt
+    if (this._common.gteHardfork('byzantium')) {
+      txReceipt = {
+        status: txRes.execResult.exceptionError ? 0 : 1, // Receipts have a 0 as status on error
+        ...abstractTxReceipt,
+      } as PostByzantiumTxReceipt
+    } else {
+      // This doesn't return the correct intermediary state root but just the
+      // initial block state root, so this just satisfies format requirements
+      // but can't be "used" in any way. Giving the correct intermediary state
+      // root would need a too depp intervention into the current checkpointing
+      // mechanism which hasn't been considered to be worth it on a HF backport
+      const stateRoot = await this.stateManager.getStateRoot(true)
+      txReceipt = {
+        stateRoot: stateRoot,
+        ...abstractTxReceipt,
+      } as PreByzantiumTxReceipt
+    }
+
```

Comment:
Ok, @s1na @alcuadrado please have a look if you are satisfied with this implementation, I thought it is not worth to update the checkpointing mechanism here, since this is a feature (which we then doesn't fully provide for pre-Byzantium) rather than consensus critical code.

We can alternatively also just fill in an empty Buffer and avoid changing the StateManager interface (might generally be a useful addition, not sure though). 🤔

Reply:
Some working note: just tested to run the blockchain tests on older HFs with `node -r ts-node/register --stack-size=1500 ./tests/tester --blockchain --fork=Byzantium`.

These are currently NOT working and throwing errors like `invalid block stateRoot` and `invalid receiptTrie`. I can confirm though that this is also happening on `master` so this is not induced by the changes here and rather need some separate investigation. @evertonfraga We should also run - I would suggest - 1 additional fork config on the blockchain tests on the nightly test runs.

The `Istanbul` tests are running ('node -r ts-node/register --stack-size=1500 ./tests/tester --blockchain --fork=Istanbul'). The older tests not running doesn't necessarily mean that this originates in one or several bugs in our code but can also have its origin on the `tests` side and eventually the older HF blockchain tests don't get accurately generated any more respectively have been at the state of the `tests`snapshot.

Reply:
I would rather suggest not to do this investigation here but do this separately. Otherwise the scope here gets too big.

Reply:
The receiptTrie field in the block header is a commitment to the tx receipts, so I wouldn't be surprised if they were incorrect. But I agree we can probably handle this separately because as you said it needs deeper changes. I'd suggest however till then to keep the StateManager interface unchanged and if the intermediate state root is incorrect anyway then just return a dummy value until we dig deeper.